### PR TITLE
Add Frontend PR Guidelines

### PR DESCRIPTION
**Pull request by chawes13**
A client requested that I create this documentation as part of our knowledge handover. After reviewing it with a few devs at LPL, we thought this might be a useful guide for LPLers as well!

This is focused exclusively on the Frontend. Additional guides could be added for Backend, Salesforce, etc., but are out of scope of this pull request.

Resolves: #19

### CODE REVIEW

File: gists/frontend-pull-request-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  - You should virtually never use `PropTypes.object` or `PropTypes.array`, instead reach for `PropTypes.shape` and `PropTypes.arrayOf`, respectively
+- If a prop is _not_ required, it should have a sensible default declared in `defaultProps`
+- Only put truly global components in `src/js/components`. If a component only applies to one subsection, keep it in that subsection's `components` folder and only move it to the global component folder if another use case for it comes up
+- Try to avoid passing React state and state setters (e.g., `x`, `setX`) directly to another component
+  - This couples the child component to the parent's implementation and typically is more so just rearranging files, rather than aligning with the React component-paradigm by creating a reusable, composable interface
+  - Reevaluate your decision to create a separate component or look for other ways to compose the functionality together (e.g., use [`children`](https://react.dev/learn/passing-props-to-a-component#passing-jsx-as-children), move state down, use the [render prop pattern](https://reactjs.org/docs/render-props.html))
```
Comment (bhennes2) [inline gists/frontend-pull-request-guidelines.md:90]: FWIW - the URL for `reactjs.org/docs` has this banner on it `These docs are old and won’t be updated. Go to [react.dev](https://react.dev/) for the new React docs.` Should we update the URL here?

**Review by bhennes2 — APPROVED** [review]
Looks good to me and is pretty comprehensive! Couldn't necessarily think of other guidelines. I did notice one of the `doc` URLs might need to be updated but that's pretty minor.

File: gists/frontend-pull-request-guidelines.md (hunk 1 above)
Comment (chawes13) [reply gists/frontend-pull-request-guidelines.md:90]: Unfortunately, there is no equivalent Render props documentation on the new docs site. They say most usages can be replaced by custom hooks, but I don't necessarily agree with that. At this point in time, I think it's still relevant to include the documentation referring to this pattern, especially given that libraries like Formik make significant use of it.

File: gists/frontend-pull-request-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+- If something is no longer being used, it's better to delete it! The more code you have, the more bugs you have
+  - This applies to files, functions, lines of code, styles, and comments. If it turns out that you need it, you can retrieve it from the Git history
+- Avoid [variable shadowing](https://en.wikipedia.org/wiki/Variable_shadowing) (i.e., using the same name for variables in different scopes). This leads to code that can be hard to read and defects that are difficult to diagnose
+- Duplicating context or sowing confusion with a variable name (e.g., using a singular `participant` variable for an array of participants)
+  - Refer to the [variable naming cheatsheet](https://github.com/kettanaito/naming-cheatsheet) for more examples
+- You probably don't need [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce). Most applications of reduce can be simplified and more performant when using [local mutation](https://typeofnan.dev/mutation-isnt-always-bad-in-javascript/)
```
Comment (jhp0621) [inline gists/frontend-pull-request-guidelines.md:76]: <a href="https://imgflip.com/i/2qnba9"></a>

File: gists/frontend-pull-request-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  - https://react.dev/learn/you-might-not-need-an-effect
+  - https://kentcdodds.com/blog/dont-sync-state-derive-it
+- Loose prop type definitions breed defects. Be as specific as you can when defining a prop type and place custom shapes in `types.js`
+  - You should virtually never use `PropTypes.object` or `PropTypes.array`, instead reach for `PropTypes.shape` and `PropTypes.arrayOf`, respectively
+- If a prop is _not_ required, it should have a sensible default declared in `defaultProps`
+- Only put truly global components in `src/js/components`. If a component only applies to one subsection, keep it in that subsection's `components` folder and only move it to the global component folder if another use case for it comes up
+- Try to avoid passing React state and state setters (e.g., `x`, `setX`) directly to another component
+  - This couples the child component to the parent's implementation and typically is more so just rearranging files, rather than aligning with the React component-paradigm by creating a reusable, composable interface
```
Comment (jhp0621) [inline gists/frontend-pull-request-guidelines.md:89]: Is there a good article for reference? Personally I find this to be one of the hardest things about React.

**Comment by chawes13** [comment]
> Concise and super helpful guidelines. Thanks for putting this together!
> 
> One thing: should dealing with "race conditions" fall under FE guidelines or is that too specific?

Hmm, could you be more specific about what you mean? It's a code smell for sure but I'm not exactly sure how to pinpoint it in the context of frontend code

File: gists/frontend-pull-request-guidelines.md (hunk 3 above)
Comment (chawes13) [reply gists/frontend-pull-request-guidelines.md:89]: Ugh I've been trying to find one forever. Maybe I should just write it

**Pull request merged**